controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -190,17 +190,18 @@ private StreamMetadataTasks(final StreamMetadataStore streamMetadataStore,
         this.requestTracker = requestTracker;
         this.retentionFrequencyMillis = new AtomicLong(Duration.ofMinutes(Config.MINIMUM_RETENTION_FREQUENCY_IN_MINUTES).toMillis());
         this.retentionClock = new AtomicReference<>(System::currentTimeMillis);
+        this.eventHelperFuture = new CompletableFuture<>();
         this.setReady();
     }
 
     @Synchronized
     public void initializeStreamWriters(final EventStreamClientFactory clientFactory,
                                         final String streamName) {
-        this.eventHelper = new EventHelper(clientFactory.createEventWriter(streamName,
+        this.eventHelperFuture.complete(new EventHelper(clientFactory.createEventWriter(streamName,
```

Comment:
This is a bad pattern. Never complete a future while holding a lock. You are holding the instance-level lock here with `@Synchronized`.

I recommend you rewrite this code to either remove the `@Synchronized` or complete this future outside of the scope of the lock.

Reply:
Changes are done as per your recommendation.